### DISCUSSION TITLE
Discussion: Project Sustainability, Helping Contributors and Upstreams

### DISCUSSION THREAD

**Discussion by mitchellh** [ghostty-org/ghostty#2350] [Feature Requests, Ideas]
Hi beta group 👋 

In the privacy of our not-so-small-and-still-rapidly-growing (800+ people at the time of writing this) beta, I want to begin a discussion on project sustainability specifically from the PoV of finances and get feedback from the community at large.

I know this is a premature discussion, but I'd rather take the time to think about this and discuss it, and I hope Ghostty will be successful to where this discussion ends up being important. 😄 

## Goals

I'll start by being clear: **my goals with Ghostty _are not financially motivated at all_**. 

I've had the privilege of having a large amount of professional success in my life and I don't plan to extract a single dollar for personal use from this project. Simultaneously, I'm more than happy to significantly fund this project from a personal basis as I already have been doing so.

But, there are other major considerations:

1. **I want contributors to be supported.** I don't expect that this project would be able to support contributors on any full time basis so that's not my intention, but it'd be great to provide _some financial assistance_ to high quality, active contributors. 

2. **I want to pay it forward to upstreams.** I'm already a big personal supporter of projects like Zig (one of the largest individual supporters), and Ghostty depends heavily on other various upstreams. If Ghostty raises awareness for those projects that's great, but if Ghostty can directly pay it forward to those projects, even better.

3. **I want to support community events.** A dream is to be able to at the very least have a meetup at an event like FOSDEM, Scale, etc. A larger dream is to be able to support a dev room at FOSDEM or something like that to talk about terminal emulation _generally_. In that case, it'd be wonderful to have a small budget for these, in particular helping high-quality, active contributors make it to these events. To reiterate above: I will always pay for my own way and I don't plan to use any of the community funds for myself personally even in my capacity as Ghostty maintainer.

## Why?

99.9% of open source projects don't bother with something like this. Ghostty is just a basic terminal emulator, so why bother worrying about this?

First, Ghostty does not _need_ to worry about this. I'm more than happy keeping Ghostty a small, personal project with the limitations that come with that. This is absolutely fine.

But my goals and ambitions for Ghostty are large: I want to create a free, fast, innovative, feature-rich, cross-platform, fully native terminal emulator. And I want it to be nearly the best in each of those categories. The Ghostty project is also comprised of **libghostty** which enables other projects to embed Ghostty, and I want to libghostty to be actively used across the ecosystem. Needless to say but I'm saying it anyways: these are big goals and the more people I can get involved, the better this can be.

**Do not fear: I have no plans to make Ghostty a for-profit corporate entity.** I've been very clear from the beginning that Ghostty will be FOSS licensed on an OSI-approved license. The license discussion is still ongoing but at the time of writing this we're leaning full MIT for the entire project. 

This project is and will continue to be first and foremost a _personal, passion project_. For the indefinite future I will be the BDFL of this project. But one day, perhaps I won't, who knows, and if I can start off from the beginning in getting this project on the best foot forward to support that possible future, that's another reason to talk about this now.

Because of the above, I think the project is better off considering project sustainability up front, even if it's putting the cart before the horse a bit. As an additional point, people already ask me a lot about this so I'd love to have a firm answer on all this.

## The Idea

My idea currently is to leverage [Polar](https://polar.sh) and provide typical project benefits:

- Your name, company highlighted in README, release notes, or website, etc.
- Access to exclusive content first (i.e. devlogs) that go public a bit later.
- Access to contributor-only discussion areas on Discord
- etc.

In addition to subscribers, I want to support some level of bountied issues or feature requests. I don't want to support open-ended bountied issues, i.e. the ability to put a bounty on _any issue_. Instead, the community could propose features like they do anyways, and if I think an issue is a good target for a bounty, we can mark it as such. Therefore, this would be a _curated list of bountied issues_.

Outside of Polar, Ghostty swag is also something I'd like to do to support the project. I already provide stickers _for free_ (out of my own pocket) for the beta community, but as the project becomes public and hopefully grows, I'd like more options and to use this as a way to support the project.

The funds would be at my discretion for how they're used (to keep things simple). But the direction they'd go:

- Contributor support
- Bountied issues
- Other project-related costs
- Upstream dependencies (i.e. Zig)
- **Never** to me personally. I'm good! 🙏

## Non-ideas

Things that are **not and never will be** on the table:

- Non-OSS licenses
- Open core business strategies
- Open-ended bountied issues (i.e. any possible issue)

Further, I want to be clear that I fully intend to keep Ghostty _my personal project_ for the indefinite future with a BDFL leadership structure. I'm not interested for the indefinite future in creating a parent ownership organization or an alternate governance structure. This isn't a "never" thing (like the above list) but it's something that isn't on the table today.

## Discussion

My question now to the beta community:

- What are your general thoughts? Yay/nay?
- Specific ideas?

**Comment by VisenDev** [comment]
I think merch and a small subscription are perfectly valid ways to support the project. This is what upstream zig is already doing for a lot of their funding (edit, the subscriptions like github sponsor or every.org, not the merch).

As far as subscriptions, I think having a really small option with some benefits would have pretty good value. Like $0.99 a month to access a special discord channel or something simple. But you could have higher gradations above that

**Comment by haze** [comment]
Is there notable prior art that you'd like to copy / take inspiration from?

If Ghostty (given its ambitious endeavors with `libghostty`) scales to the likes of Emacs and Python, will there be challenges / changes in the support model that you foresee? 

I'm happy with and approve of the current selection of contributions and donation mechanisms.

**Comment by qwerasd205** [comment]
I'll start off by saying that the commitment to keeping Ghostty a passion project and fully FOSS is a relieving one. I feel like I've seen many excellent open source projects succumb to a path of open-core followed by inevitable enshittification, so assurance that that won't ever happen to Ghostty is good to have (and significantly bolsters my willingness to spend time and energy contributing).

Your listed Polar benefits sound very acceptable, specifically in the fact that they don't create a paywall for anything that the general public or userbase would have a vested interest in access to.

A curated list of bountied issues makes a lot of sense to me as well, allowing you to steer the project in a singular focused direction ultimately decided by you as the BDFL.

Supporting the upstream is really important! I think placing an emphasis on it is definitely the right choice; too many FOSS projects are completely disrespected and left unsupported by their downstreams and the culture around that absolutely needs to change, so any step towards normalizing the support of your upstream is a good one in my book.

I don't think establishing things like this early on is jumping the gun or putting the cart before the horse; the longer you wait the harder it gets as more people become involved with their own unchecked assumptions that they brought with them. Making it clear what the terms of the project are and how things will happen lets people make better informed decisions on if they want to invest their time and energy in to contributing or if they'd rather go somewhere else.

So to answer the specific questions in the original post:
- My general thoughts? See above.
- Yay/nay? Absolutely yay.
- Specific ideas? I can't say I have any at this time.

**Comment by jcollie** [comment]
I'm in favor. I'd definitely be down for some quality merch like t-shirts or baseball caps. A dark blue/purple baseball cap with an embroidered Ghostty logo would be killer. I'd be willing to kick in a couple of bucks a month to help pay for CI and other hosting costs as well.

**Comment by VisenDev** [comment]
I think some Tshirts with art of the Ghostty Ghost beating-up/fighting emulators like PowerShell or iterm2 would be pretty funny

**Comment by andrewrk** [comment]
Love this. Sustainability is an important meta-feature for any open source project.

Here's my analysis:

## Long-Term Sustainability

Observation: the license acts as a canary, but realistically, the project is made of the active contributors who work on the ongoing maintenance of the project.

One example would be freenode. The leadership did something shitty, I can't even remember what it was tbh, but the actual people doing the work to keep the network alive all rebelled and started libera.chat. Now freenode is basically dead and everybody moved over to libera.chat.

Another hypothetical example would be: if Mitchell retired (i.e. stopped contributing code) and Amazon hired all the main contributors of ghostty, then Amazon is now in a position where it can change the license and get away with it. The users would follow the activity from the main, established contributors of the project. I'm sure there are exceptions to this, but the purpose of this analysis is to prepare for likely scenarios.

Typical risk points for a company or project taking a turn for the worse would be:
- [when the company owner or BDFL retires](https://techcrunch.com/2021/01/07/epic-acquires-rad-game-tools-veteran-of-many-gaming-generations/)
- when the company owner or BDFL needs money for their children to go to college

It seems like only the first point applies in this case. It also looks like it will probably be a long time before Mitchell wants to retire. But it would be worth coming up with a plan for project longevity specifically when that event starts to approach. Until that point, it seems low risk to me.

There is one more important factor here, which is something I don't see many people thinking about explicitly: modeling your dependencies, not just as code, but more specifically as labor liabilities.

Mitchell is already instinctively doing this as he mentioned above with the idea of "paying it forward to upstreams". To illustrate this idea, let's look at the example of zlib, which sees extremely low churn. There's what, a release every 10 years now? It requires little to no labor to have a zlib dependency. It's almost free.

On the other hand, depending on Zig requires considerably more labor due to language churn and standard library API churn. Despite causing plenty of breaking changes myself, I'm quite aware of the consequences to downstream projects. A pre-1.0 Zig dependency is a *labor liability*.

Labor liabilities make the project cost more. It literally requires labor to keep the project from bitrotting, and labor requires feeding, housing, and providing medical care for human beings. That costs money.

My point is that projects can and should think carefully about their dependencies, not only from a software engineering perspective, but a human labor perspective. By limiting the amount of ongoing maintenance work that must occur to keep a project alive, the project costs less money to maintain and therefore becomes more sustainable. Careful use of this strategy can even result in a project being so cheap to maintain that a single hobbiest can provide those services in their spare time, without compromising health or happiness!

Put this way, it's quite a compliment that Mitchell has chosen to use Zig for Ghostty despite these tradeoffs, and I'm honored. I will of course do my best to reach the 1.0 milestone so that Zig can minimize its ongoing maintenance cost as a dependency.

More examples of high-cost maintenance burdens:
* Integration with a third-party API such as Google, Twitter, Facebook, Spotify, etc.
* Relying on web page scraping
* Supporting Apple operating systems ([haha only serious](https://wiki.c2.com/?HaHaOnlySerious))

I'm not saying don't do these things ever, obviously, but it's worth modeling them explicitly and knowing that you're paying a recurring human labor cost to keep the features that depend on them alive.

## Polar as a Revenue Platform

I have not personally used Polar, nor have I heard of it before today. Here are my initial observations:

It is a for-profit company, operating on a platform capitalism model. For those who are not already cautious about dealing with such things, I recommend this talk:

[DEF CON 31 - An Audacious Plan to Halt the Internet's Ensh*ttification - Cory Doctorow](https://www.youtube.com/watch?v=rimtaSgGz_4)

In summary, the danger is that, when the platform is small (i.e. right now), it spends venture capital money to offer an enticing service to consumers. However, as it gains mainstream appeal, the company is strongly incentivized to squeeze more value out of the platform, and consumers are stuck to the platform due to network effects. Once is not enough however; it must make *more* profit every year, so it will continue to squeeze the value from the platform up to and including the point at which it destroys the platform and everybody has to leave. For example, if ghostty was getting 90% of its revenue from Polar, then one day Polar will probably decide to increase its revenue by adjusting the 5% percent revenue share to a bigger number, and ghostty will have no recourse.

This pattern is extremely strong in American companies but weaker in Swedish companies, I think. So that's a positive signal.

The problem is that there just aren't many revenue platform options.

For Zig Software Foundation, I was pleased to discover Every.org, which is itself a non-profit organization dedicated to fundraising for other non-profits. Ultimately, every organization is made up of people, and every set of people have their own strengths and weaknesses, however, at least with a non-profit you don't have a strong incentive pulling the organization in a terrible direction. Every.org would be an option if there was a 501(c)(3) organization corresponding to ghostty, but not otherwise.

In conclusion, I think my suggestion would be to go ahead and use Polar if it seems useful, but maybe also try to diversify across multiple revenue platforms. This would at least give ghostty some leverage if any of the revenue platforms started doing any funny business. And expect to have to migrate off the platform in 5-10 years.

## Merch

> I think merch and a small subscription are perfectly valid ways to support the project. This is what upstream zig is already doing for a lot of their funding.

Just to make sure nobody misunderstands, [merch is approximately 0% of Zig Software Foundation's income](https://ziglang.org/news/2024-financials/#2023-income). From my perspective, merch is actually a service a project provides to the community that is net-negative in terms of income/labor. Consider that you don't want to purchase a new sticker every time a commit lands in ziglang/zig. That said, it would be amusing to see a laptop with 28,686 stickers on it!

**Comment by mitchellh** [comment]
Just responding to a few pointed things...

> If Ghostty (given its ambitious endeavors with libghostty) scales to the likes of Emacs and Python, will there be challenges / changes in the support model that you foresee?

I think taken to the limit, any good plan falls apart and has to be adjusted. If Ghostty reaches this level of success, then we'll have difficulties no matter what. I don't think the plan in this issue will scale to that level, but nor do I think it has to. The goal is to give us a good starting point to iterate to that if the situation arises.

> In conclusion, I think my suggestion would be to go ahead and use Polar if it seems useful, but maybe also try to diversify across multiple revenue platforms.

I agree. I'd like to start with Polar (so far) but I have every intention to look at additional platforms. I think I can use the [Polar API](https://api.polar.sh/redoc) so that the rewards can be centrally managed by Polar while money comes from additional places too, as a 2nd step (1st step being simply adopting Polar full out). 

> From my perspective, merch is actually a service a project provides to the community that is net-negative in terms of income/labor. 

I agree that merch is high work for low direct financial return. However, I think the real _financial_ value in merch is that it is excellent, cheap marketing. If someone isn't willing to subscribe, but they're willing to buy and wear a t-shirt, there is a small chance they make someone else aware of the project who _does_ subscribe. 

So, in addition to being fun and rewarding for people to have branded swag, for the project itself I think merch is a great way to simply raise awareness for the other, more real sustainability options.

**Comment by kristoff-it** [comment]
As mentioned in previous comments, depending on how Ghostty ends up being hosted fiscally, different platforms will make more or less sense for it.

I would like to bring up [OpenCollective](https://opencollective.com/) as an option because it offers a potentially interesting middle ground, where the project can be hosted by another non-profit entity. The main reason we're not using it for Zig is because we have our own fiscal host already.

This means that, at least on paper, one could get all the benefits from being a non-profit, without incurring in any of the administrative costs of running one. 

Of course, in reality, your fiscal host at that point would be a dependency of some kind, with similar downsides to the ones mentioned by Andrew wrt software dependencies.

That said, I think this point is relevant also because it segues into the topic of funding sources.

I think we can all agree that, at least in principle, those who benefit from Ghostty are the ones who should be contributing financially according to their means.

Based on my experience at the ZSF, this tends to boil down to two main aspects:

1. The project voicing a need for money.
2. The project being present in the user's mind.

The first one is obvious: if the project doesn't say that it needs money, people will not go out of their way to investigate if that could be the case. Some will donate regardless out of gratitude, but it is in general a good idea to let the public know that money is needed and for what general purpose. From that perspective, being a non profit helps because it both makes your finances public and it also gives people a higher level of confidence that the information is 100% genuine.

The second point is also obvious in its own way: people donate only to projects they *care* about, and to care they need to be made part of the journey. A journey requires an active pursuit of a mission, which might or might not be compatible with what Mitchell wants to do with Ghostty.

For example one perfectly reasonable way to run the project would be to just focus on coding a good terminal emulator and not do any communication, but at that point I would not be surprised if over time people stop *caring* about Ghostty and the project ends up not being able to send contributors to FOSDEM using donation money anymore, even though the project would be a success by the stated metric.

Unfortunately some kinds of success are incompatible with long-term sustainability and once people just take for granted the software, then their interest will also be directed somewhere else. I do think that this is a puzzle worth solving because it would be nice to consider a project *done* eventually but, going back to the main topic, my point is that if you want to have the kind of money that can pay for PRs and send contributors to conferences, then you will most probably need a story around your project.

Considering that Mitchell expressed an interest in discussing with other terminal emulator projects the creation of new standards and in general improving the emulator ecosystem, it might be worth considering going "all-in" with that idea as that would be a journey long and compelling enough to keep people interested in the project for a good while, ideally for long enough to allow Ghostty to transition to maintenance mode.

**Comment by mitchellh** [comment]
Thanks @kristoff-it for your input!

I think your points about having a mission (and being vocal about it) make sense. My current thinking is that the primary mission is going to be libghostty and that downstream that utilizes libghostty (in particular the commercial entities) can be encouraged/pressured to help support the upstream project. 

The new standards and improving terminal emulation is something I am interested in, but I think it'd be a tough thing to jump right into without building more of a reputation of being a good citizen in the general TE community. I'm still thinking about the right way to approach this...

**Reply by jotihojr** [reply]
As pertaining to `libghostty`...
Without knowing the practical implications of and the real commercial viability with `libghostty`, I believe a minimal non-profit institution holding semi-restrictive, minimal cost licensing for bulk commercial use of derivative utilities (i.e. something in the form of  a SSH client) could provide for _perpetual funding_.

**Comment by mitchellh** [comment]
FYI, relevant to this issue: https://mitchellh.com/writing/polar

This is a bias, but I want to note that my original thoughts around this post **_predate_** ever meeting with or committing to being involved with Polar. On the flip side, one of the motivating factors to be more directly involved with Polar was in the scenario that Ghostty uses it 😄 That isn't a certainty yet and as the blog post above states, it aligns with my personal interests regardless, but I want to be totally transparent here.

**Comment by g-cassie** [comment]
I think you are in a very unique position to chart a new path for an open source project and it's exciting to see you're thinking about this so proactively.  It also seems like there is a reasonable amount of agreement here that establishing some source of funding would be beneficial for the project. I wanted to share a few of my thoughts on this topic which I don't think are reflected in the thread yet.

### How should Ghostty make money?

With respect to how Ghostty makes money, I would consider a more conventional product revenue stream.  Rather than asking for a monthly donation with various benefits, ask people to purchase a license to use Ghostty (either because legally required for "commercial use" or simply for "good karma").  Developers working in large tech companies can expense a $20-50 license for a tool with little justification. On the other hand, a $x/month  donation to an open source community is not so easy to expense. If Ghostty required a small payment for a perpetual (one year?) license I think it would meet its fundraising goals faster than with donations - and at a much lower personal cost to its users. 

This is kind of a hybrid with a donation model because with Ghostty being firmly OSS, anyone could easily compile their own binary of Ghostty and use it freely. It's essentially extremely low friction nagware. I also think it's somewhat appropriate that an organization spends millions/billions in annual software expenditures and uses Ghostty, a small sliver of that should end up supporting the project. 

The subscription donation model of open source funding is a series of compromises to address the fact that it's very hard to put a price on a library or framework which is used by a niche group of programmers.  As an end-user product which will have thousands of every day professional users, Ghostty does not have these problems.  Subscription revenue is fickle and large donors can exert significant influence.  It would be nice for the core contributors of Ghostty to maintain the ability to build their product in an opinionated way.  You can imagine a large sponsorship from Apple or Microsoft could quickly shift priorities. Revenue from thousands of customers will do the best to preserve this.  It may even be the case that revenue exceeds expenses and a long term endowment can be created for the future of Ghostty.

### Who collects the money?

As a few commenters have highlighted, once money comes in it needs to go somewhere.   This opens Pandora's box on all the types of legal entities and structures that can be considered.  Not only are these difficult questions with long reaching implications, it also just means a lot of non-technical work which is not what most open source maintainers want to do.  I can see why something like Polar or OpenCollective might be appealing to avoid this.  At some point though, you are still forced to make decisions about governance (who controls where the money goes next) - and having legal structure in place could make that easier to navigate - in which case your legal entity can just open a bank account and pay people and all that.

### What license?

A highly permissive license like MIT will knee cap many of the adverse outcomes that have been mentioned above - as demonstrated recently with redis.  It's a double edge sword though - a Microsoft backed fork of Ghostty that ships on fresh installs of Windows could quickly snuff out a humble open source organization.

**Comment by eljobe** [comment]
I just wanted to drop into this discussion and indicate my strong support for using the MIT license when we go public with ghostty. I know there is a real cooling effect for projects who use less permissive licenses.

**Reply by jcollie** [reply]
I'm pretty sure that's the plan. Mitchell has asked anyone that's previously contributed code to Ghostty permission to re-license their code to MIT.
